locator/topology.hh

```diff
@@ -41,6 +41,35 @@ using node_holder = std::unique_ptr<node>;
 
 using shard_id = seastar::shard_id;
 
+class datacenter;
+class rack;
+
+class location {
+    const datacenter* _dc = nullptr;
+    const rack* _rack = nullptr;
+
+public:
+    location() = default;
+    location(const datacenter* dc, const rack* rack) noexcept : _dc(dc), _rack(rack) {}
```

Comment:
I think it's unsafe to make it a value object and hold unprotected pointers to _dc and _rack. Users may think they're copying location and pass it outside the scope of token_metadata_ptr.

Reply:
What would you suggest?
To remove the method and to just have node::dc() and node::rack()
Or return datacenter_id and rack_id wrapped in location instead?

Reply:
node::dc() and node::rack() are a good idea regardless.

We could also have location() which returns location* and make location non-copyable.

Reply:
Please see https://github.com/scylladb/scylladb/pull/17273#issuecomment-1943316121 / #17336 for an alternative approach for the dc/rack identifiers.